crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+) {
+    if let [expr] = targets.as_slice() {
+        sort_dunder_all(checker, expr, value);
+    }
+}
+
+/// Sort an `__all__` mutation represented by a `StmtAugAssign` AST node.
+/// For example: `__all__ += ["b", "c", "a"]`.
+pub(crate) fn sort_dunder_all_aug_assign(checker: &mut Checker, node: &ast::StmtAugAssign) {
+    if node.op.is_add() {
+        sort_dunder_all(checker, &node.target, &node.value);
+    }
+}
+
+/// Sort an `__all__` mutation from a call to `.extend()`.
+pub(crate) fn sort_dunder_all_extend_call(
+    checker: &mut Checker,
+    ast::ExprCall {
+        func,
+        arguments: ast::Arguments { args, keywords, .. },
+        ..
+    }: &ast::ExprCall,
+) {
+    let ([value_passed], []) = (args.as_slice(), keywords.as_slice()) else {
+        return;
+    };
+    let ast::Expr::Attribute(ast::ExprAttribute {
+        ref value,
+        ref attr,
+        ..
+    }) = **func
+    else {
+        return;
+    };
+    if attr == "extend" {
+        sort_dunder_all(checker, value, value_passed);
+    }
+}
+
+/// Sort an `__all__` definition represented by a `StmtAnnAssign` AST node.
+/// For example: `__all__: list[str] = ["b", "c", "a"]`.
+pub(crate) fn sort_dunder_all_ann_assign(checker: &mut Checker, node: &ast::StmtAnnAssign) {
+    if let Some(value) = &node.value {
+        sort_dunder_all(checker, &node.target, value);
+    }
+}
+
+fn sort_dunder_all(checker: &mut Checker, target: &ast::Expr, node: &ast::Expr) {
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target else {
+        return;
+    };
+
+    if id != "__all__" {
+        return;
+    }
+
+    // We're only interested in `__all__` in the global scope
+    if !checker.semantic().current_scope().kind.is_module() {
+        return;
+    }
+
+    let (elts, range, kind) = match node {
+        ast::Expr::List(ast::ExprList { elts, range, .. }) => (elts, *range, DunderAllKind::List),
+        ast::Expr::Tuple(tuple_node @ ast::ExprTuple { elts, range, .. }) => {
+            (elts, *range, DunderAllKind::Tuple(tuple_node))
+        }
+        _ => return,
+    };
+
+    let mut possibly_fixable = true;
+    let mut string_items = vec![];
+    for elt in elts {
+        // Don't flag `__all__` definitions that contain non-strings
+        let Some(string_literal) = elt.as_string_literal_expr() else {
+            return;
+        };
+        // If any strings are implicitly concatenated, don't bother trying to autofix
+        if string_literal.value.is_implicit_concatenated() {
+            possibly_fixable = false;
+        }
+        string_items.push(string_literal.value.to_str());
+    }
+    if dunder_all_is_already_sorted(&string_items) {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);
+
+    if possibly_fixable {
+        if let Some(fix) = get_fix(range, elts, &string_items, &kind, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
```

Comment:
It's hard to tell if it's worth it (probably not because `__all__` items are so uncommon). But we could avoid allocating a new `Vec` when the items are already sorted.


```suggestion
    let Some(unsorted) = UnsortedDunderAllItems::from_elements(elts) else {
        return;
    };

    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);

    if !unsorted.implicit_concatenated {
        if let Some(fix) = get_fix(range, elts, &unsorted.items, &kind, checker) {
            diagnostic.set_fix(fix);
        }
    }

    checker.diagnostics.push(diagnostic);
}

struct UnsortedDunderAllItems<'a> {
    items: Vec<&'a str>,
    implicit_concatenated: bool,
}

impl<'a> UnsortedDunderAllItems<'a> {
    fn from_elements(elements: &'a [Expr]) -> Option<Self> {
        let Some((first, rest)) = elements.split_first() else {
            return None;
        };

        if rest.is_empty() {
            return None;
        }

        let mut previous = first.as_string_literal_expr()?.value.to_str();

        for element in rest {
            let current = element.as_string_literal_expr()?.value.to_str();

            if AllItemSortKey::from(current) < AllItemSortKey::from(previous) {
                let mut implicit_concatenated = false;

                let mut items = Vec::with_capacity(elements.len());

                for element in elements {
                    let string_literal = element.as_string_literal_expr()?;
                    implicit_concatenated |= string_literal.value.is_implicit_concatenated();
                    items.push(string_literal.value.to_str());
                }

                return Some(Self {
                    items,
                    implicit_concatenated,
                });
            }

            previous = current;
        }

        // All Sorted
        None
    }
}

```